examples/cpp/7_series/gnss_ins/7_series_gnss_ins_example.cpp

```diff
@@ -0,0 +1,1194 @@
+            SAMPLE_RATE_HZ,
+            gnssBaseRate2
+        );
+    }
+
+    // Calculate the decimation (stream rate) for the device based on its base rate
+    const uint16_t gnssDecimation2 = gnssBaseRate2 / SAMPLE_RATE_HZ;
+    MICROSTRAIN_LOG_INFO("Decimating GNSS 2 base rate %d by %d to stream data at %dHz.\n",
+        gnssBaseRate2,
+        gnssDecimation2,
+        SAMPLE_RATE_HZ
+    );
+
+    // Descriptor rate is a pair of data descriptor set and decimation
+    const mip::DescriptorRate gnssDescriptors2[1] = {
+        { mip::data_gnss::FixInfo::FIELD_DESCRIPTOR, gnssDecimation2 }
+    };
+
+    MICROSTRAIN_LOG_INFO("Configuring message format for GNSS 2 data.\n");
+    cmdResult = mip::commands_3dm::writeMessageFormat(
+        _device,
+        mip::data_gnss::MIP_GNSS2_DATA_DESC_SET,                // Data descriptor set
+        sizeof(gnssDescriptors2) / sizeof(gnssDescriptors2[0]), // Number of descriptors to include
+        gnssDescriptors2                                        // Descriptor array
+    );
+
+    if (!cmdResult.isAck())
+    {
+        terminate(_device, cmdResult, "Could not set message format for GNSS 2 data!\n");
+    }
+}
+
+////////////////////////////////////////////////////////////////////////////////
+/// @brief Configures message format for filter data streaming
+///
+/// @details Sets up filter data output by:
+///          1. Querying device base rate
+///          2. Validating desired sample rate against base rate
+///          3. Calculating proper decimation
+///          4. Configuring message format with:
+///             - GPS time
+///             - Filter status
+///             - LLH position
+///             - NED velocity
+///             - Euler angles
+///
+/// @param _device Reference to the initialized MIP device interface
+///
+void configureFilterMessageFormat(mip::Interface& _device)
+{
+    // Note: Querying the device base rate is only one way to calculate the descriptor decimation
+    // We could have also set it directly with information from the datasheet
+
+    MICROSTRAIN_LOG_INFO("Getting the base rate for filter data.\n");
+    uint16_t       filterBaseRate;
+    mip::CmdResult cmdResult = mip::commands_3dm::getBaseRate(
+        _device,
+        mip::data_filter::DESCRIPTOR_SET, // Data descriptor set
+        &filterBaseRate                   // Base rate out
+    );
+
+    if (!cmdResult.isAck())
+    {
+        terminate(_device, cmdResult, "Could not get filter base rate!\n");
+    }
+
+    // Supported sample rates can be any value from 1 up to the base rate
+    // Note: Decimation can be anything from 1 to 65,565 (uint16_t::max)
+    if (SAMPLE_RATE_HZ == 0 || SAMPLE_RATE_HZ > filterBaseRate)
+    {
+        terminate(
+            _device,
+            mip::CmdResult::NACK_INVALID_PARAM,
+            "Invalid sample rate of %dHz! Supported rates are [1, %d].\n",
+            SAMPLE_RATE_HZ,
+            filterBaseRate
+        );
+    }
```

Comment:
Supported rates are technically filterBaseRate/N, where N is an integer. E.g. you could change the logic to `(filterBaseRate/SAMPLE_RATE_HZ*SAMPLE_RATE_HZ != filterBaseRate)`